src/invidious/jsonfilter.cr

```diff
@@ -0,0 +1,251 @@
+        closing_bracket_index = bracket_pairs[opening_bracket_index]
+        content_start = opening_bracket_index + 1
+
+        content = fields_text[content_start...closing_bracket_index]
+
+        if content.empty?
+          raise FieldsParser::ParseError.new "Empty nest group at position #{content_start}"
+        else
+          content = remove_nest_groups(content)
+        end
+
+        while nest_stack.size > 0 && closing_bracket_index > nest_stack[nest_stack.size - 1][:closing_bracket_index]
+          if nest_stack.size
+            nest_stack.pop
+          end
+        end
+
+        group_name.split('/').each do |group_name|
+          nest_stack.push({
+            group_name:            group_name,
+            closing_bracket_index: closing_bracket_index,
+          })
+        end
+
+        if !content.empty?
+          properties = content.split(',')
+
+          properties.each do |prop|
+            nest_list = nest_stack.map { |nest_prop| nest_prop[:group_name] }
+
+            if !prop.empty?
+              if prop.includes?('/')
+                parse_single_nests(prop) { |list| nest_list += list }
+              else
+                nest_list.push prop
+              end
+            else
+              raise FieldsParser::ParseError.new "Empty key in nest list: #{nest_list << prop}"
+            end
+
+            yield nest_list
+          end
+        end
+      end
+    end
+
+    def self.remove_nest_groups(text : String) : String
+      content_bracket_pairs = get_bracket_pairs(text, false)
+
+      content_bracket_pairs.each_key.to_a.reverse.each do |opening_bracket|
+        closing_bracket = content_bracket_pairs[opening_bracket]
+        last_comma = text.rindex(',', opening_bracket) || 0
+
+        text = text[0...last_comma] + text[closing_bracket + 1...text.size]
+      end
+
+      return text.starts_with?(',') ? text[1...text.size] : text
+    end
+
+    def self.get_bracket_pairs(text : String, recursive = true) : BracketIndex
+      istart = [] of Int64
+      d = BracketIndex.new
+
+      text.each_char_with_index do |char, index|
+        if char == '('
+          istart.push(index.to_i64)
+        end
+
+        if char == ')'
+          begin
+            opening = istart.pop
+            if recursive || (!recursive && istart.size == 0)
+              d[opening] = index.to_i64
+            end
+          rescue
+            raise FieldsParser::ParseError.new "No matching opening parenthesis at: #{index}"
+          end
+        end
+      end
+
+      if istart.size != 0
+        idx = istart.pop
+        raise FieldsParser::ParseError.new "No matching closing parenthesis at: #{idx}"
+      end
+
+      return d
+    end
+  end
+
+  class FieldsGrouper
+    alias JSONSkeleton = Hash(String, JSONSkeleton)
+    alias SkeletonValue = Hash(String, JSONSkeleton)
+
+    def self.create_json_skeleton(fields_text : String) : SkeletonValue
+      root_hash = SkeletonValue.new
```

Comment:
The reason `SkeletonValue.new` doesn't work using the alias I mentioned above is because the compiler needs to know if the variable should be decalred as a `String` or `Hash`.

Instead:
```crystal
root_hash = {} of String => SkeletonValue
```